### PR TITLE
feat: Initial commit for REST commands functionality update

### PR DESCRIPTION
**Pull request by sudhamani-hcl**
Signed-off-by: Sudhamani Bijivemula <sudhamani.b@hcl.com>

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/device-rest-go/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [x] I am not introducing a new dependency (add notes below if you are)
- [ ] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [ ] I have opened a PR for the related docs change (if not, why?)

## Testing Instructions
Testing procedure is mentioned in the README.md file under `Testing/Simulation of REST client` section

### CODE REVIEW

**Comment by cloudxxx8** [comment]
@sudhamani-hcl could you please open issues first to describe what you want to modify in this service?  The different issues should be opened for different features or fixes.

**Comment by cloudxxx8** [comment]
since we just upgraded the modules to v3, please help rebase and resolve the conflicts.

File: driver/restdriver.go [hunk 1]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+		// GET response header.  Take response body as it is if device
+		// resource data type is binary or object. For other data types convert
+		// response body to string to use during validation of reponse
+		if deviceResource.Properties.ValueType == common.ValueTypeBinary ||
+			deviceResource.Properties.ValueType == common.ValueTypeObject {
+			reading = body
+		} else {
+			reading = string(body)
+		}
+		contentType := resp.Header.Get("Content-Type")
```
Comment (cloudxxx8) [inline driver/restdriver.go:303]: use https://github.com/edgexfoundry/go-mod-core-contracts/blob/7e0c83d7b629306e64997ecd877af1515e598311/common/constants.go#L289

File: driver/restdriver.go [hunk 2]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+			}
+
+			// Create new PUT request, this will not send request to end device
+			request, err = http.NewRequest(http.MethodPut, uri, bytes.NewReader(buf))
+			if err != nil {
+				// handle error
+				return fmt.Errorf("PUT request creation failed")
+			}
+			// Set content type as application/json
+			request.Header.Set("Content-Type", "application/json; charset=UTF-8")
```
Comment (cloudxxx8) [inline driver/restdriver.go:419]: use 
https://github.com/edgexfoundry/go-mod-core-contracts/blob/7e0c83d7b629306e64997ecd877af1515e598311/common/constants.go#L289
https://github.com/edgexfoundry/go-mod-core-contracts/blob/7e0c83d7b629306e64997ecd877af1515e598311/common/constants.go#L292

File: driver/restdriver.go [hunk 3]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+				return fmt.Errorf("PUT request data is not valid")
+			}
+			// Create new PUT request
+			request, err = http.NewRequest(http.MethodPut, uri, strings.NewReader(cast.ToString(reading)))
+			if err != nil {
+				// handle error
+				return fmt.Errorf("PUT request creation failed")
+			}
+			// Set content type as text/plain
+			request.Header.Set("Content-Type", "text/plain; charset=UTF-8")
```
Comment (cloudxxx8) [inline driver/restdriver.go:439]: use 
https://github.com/edgexfoundry/go-mod-core-contracts/blob/7e0c83d7b629306e64997ecd877af1515e598311/common/constants.go#L289
https://github.com/edgexfoundry/go-mod-core-contracts/blob/7e0c83d7b629306e64997ecd877af1515e598311/common/constants.go#L295

File: driver/restdriver.go [hunk 4]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+		if !ok {
+			driver.logger.Errorf("Incoming Writing ignored. Resource '%s' not found", resourceName)
+			return nil
+		}
+
+		// Now get the query parameters received in the request.
+		// These needs to be sent to the end device
+		// Query parameters needs to be converted to string to append to the uri
+		driver.logger.Debugf("URLRawQuery :%v", req.Attributes["urlRawQuery"])
+		reqParam := fmt.Sprint(req.Attributes["urlRawQuery"])
```
Comment (cloudxxx8) [inline driver/restdriver.go:386]: claim `"urlRawQuery"` as a constant

**Review by cloudxxx8 — CHANGES_REQUESTED** [review]

File: cmd/res/devices/ED_ID1.toml [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  [DeviceList.Protocols]
+    [DeviceList.Protocols.EndDevice_Params]
+	  ED_IP = "127.0.0.1"
+	  ED_PORT = "5000"
+	  #ED_URI_Prefix = ""
+	  ED_URI_Prefix = "api"
+  [[DeviceList.AutoEvents]]
+    Interval = "20s"
+    OnChange = false
+    SourceName = "jsonRes"
```
[the remark this answers is not in the document]
Comment (sudhamani-hcl) [reply cmd/res/devices/ED_ID1.toml]: Commented as suggested

File: driver/protocolpropertykey.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+// 2. Sudhamani Bijivemula   HCL Technologies
+// 3. Vijay Annamalaisamy    HCL Technologies
+//
+
+package driver
+
+const (
+	ED_IP   = "ED_IP"
+	ED_PORT = "ED_PORT"
+	// ED_URI_PREFIX = ""
```
[the remark this answers is not in the document]
Comment (sudhamani-hcl) [reply driver/protocolpropertykey.go]: Removed the commented line

File: driver/restdriver.go [hunk 7]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+		if !ok {
+			driver.logger.Errorf("Incoming reading ignored. Resource '%s' not found", resourceName)
+			return nil, fmt.Errorf("Resource not found")
+		}
+
+		// Now get the query parameters received in the request.
+		// These needs to be sent to the end device
+		// Query parameters needs to be converted to string to append to the uri
+		driver.logger.Debugf("URLRawQuery :%v", req.Attributes["urlRawQuery"])
+		reqParam := fmt.Sprint(req.Attributes["urlRawQuery"])
```
[the remark this answers is not in the document]
Comment (sudhamani-hcl) [reply driver/restdriver.go]: Defined as constant as suggested

File: driver/restdriver.go (hunk 1 above)
Comment (sudhamani-hcl) [reply driver/restdriver.go]: Using the constants as suggested

File: driver/restdriver.go (hunk 3 above)
Comment (sudhamani-hcl) [reply driver/restdriver.go]: Using the constants as suggested

File: driver/restdriver.go (hunk 4 above)
Comment (sudhamani-hcl) [reply driver/restdriver.go]: Defined as constant as suggested

**Comment by sudhamani-hcl** [comment]
Hi @cloudxxx8,

I have addressed all the review comments. Could you please review and let me know your comments.

Thanks,
Sudhamani

File: cmd/res/profiles/ED_ID1.yaml [hunk 8]
```diff
@@ -0,0 +1,114 @@
+name: "ED_ID1"
```
Comment (lenny-intel) [inline cmd/res/profiles/ED_ID1.yaml:1]: Also change the file name accordingly 
```suggestion
name: "sample-2way-rest-device"
```

File: cmd/res/profiles/ED_ID1.yaml [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+manufacturer: "HCL Corp."
+model: "Some 3rd party app sending data"
+labels:
+ - "rest"
+ - "ED_ID1"
```
Comment (lenny-intel) [inline cmd/res/profiles/ED_ID1.yaml:6]: Remove.

File: cmd/res/profiles/ED_ID1.yaml [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+description: "REST Device"
```
Comment (lenny-intel) [inline cmd/res/profiles/ED_ID1.yaml:7]:
```suggestion
description: "REST Devices that handle two-way comunication"
```

File: cmd/res/profiles/ED_ID1.yaml [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+deviceResources:
+  - name: jsonRes
```
Comment (lenny-intel) [inline cmd/res/profiles/ED_ID1.yaml:10]: Please remove `Res` from all the resource names
```suggestion
  - name: json
```

File: cmd/res/profiles/ED_ID1.yaml [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    isHidden: false
+    description: "Test resource 1 Message"
+    properties:
+      valueType: "Object"
+      readWrite: "RW"
+      mediaType: "application/json"
+  - name: stringRes
```
Comment (lenny-intel) [inline cmd/res/profiles/ED_ID1.yaml:17]:
```suggestion
  - name: device-info
```

File: driver/protocolpropertykey.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+//
+
+package driver
+
+const (
+	ED_IP         = "ED_IP"
+	ED_PORT       = "ED_PORT"
+	ED_URI_PREFIX = "ED_URI_Prefix"
+	ED_PARAMS     = "EndDevice_Params"
+	URLRawQuery   = "urlRawQuery"
```
Comment (lenny-intel) [inline driver/protocolpropertykey.go:22]:
```suggestion
	RESTHost= "Host"
	RESTPort       = "Port"
	RESTPath= "Path"
	RESTProtocol     = "REST"
	URLRawQuery   = "urlRawQuery"
```

File: driver/restdriver.go [hunk 14]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+func (driver *RestDriver) HandleReadCommands(deviceName string, protocols map[string]models.ProtocolProperties, reqs []dsModels.CommandRequest) (responses []*dsModels.CommandValue, err error) {
+
+	driver.logger.Debugf("In HandleReadCommands function")
+	var reading interface{}
+	var val interface{}
+	var result = &dsModels.CommandValue{}
+	var uri string
+
+	// Declare response structure,
+	// This will be filled with response received from HTTP GET request
```
Comment (lenny-intel) [inline driver/restdriver.go:192]: Please limit comments to `why` something is being done that can not be figured out be reading the code.  i.e. this line doesn't need a comment. Please apply this to all other comments.

File: driver/restdriver.go [hunk 15]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+	if !paramsExists {
+		return nil, fmt.Errorf("No End device parameters defined in the protocol list")
+	}
+
+	// Get end device IP address
+	address, ok := protocolParams[ED_IP]
+	if !ok {
+		return nil, fmt.Errorf("ED_IP not found")
+	}
+	driver.logger.Debugf("ED_IP:%v", address)
```
Comment (lenny-intel) [inline driver/restdriver.go:208]: Please wait to end and do one debug statement with all the values.

File: driver/restdriver.go [hunk 16]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+	driver.logger.Debugf("ED_URI_PREFIX:%v", uriPrefix)
+
+	// Now, we have got required end device information, its time to create GET request
+	for i, req := range reqs {
+		// First get device resource instance, needed during validation of the
+		// response data received from the end device later.
+		// RunningService returns the Service instance which is running.
+		// service.DeviceResource retrieves the specific DeviceResource instance
+		// from cache according to the Device name and Device Resource name
+		resourceName := req.DeviceResourceName
```
Comment (lenny-intel) [inline driver/restdriver.go:232]: Why create variable when you can just use req.DeviceResourceName?

File: driver/restdriver.go [hunk 17]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+		// Now get the query parameters received in the request.
+		// These needs to be sent to the end device
+		// Query parameters needs to be converted to string to append to the uri
+		driver.logger.Debugf("URLRawQuery :%v", req.Attributes[URLRawQuery])
+		reqParam := fmt.Sprint(req.Attributes[URLRawQuery])
+
+		// Form URI from the end device parameters and request parameters and
+		// query parameters. Omit uri prefix if it is empty
+		if uriPrefix != "" {
+			uri = "http://" + address + ":" + port + "/" + uriPrefix + "/" + req.DeviceResourceName + "?" + reqParam + ""
```
Comment (lenny-intel) [inline driver/restdriver.go:251]: Use fmt.Sprintf to build the uri's

File: driver/restdriver.go [hunk 18]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+
+		// Reached here, as the success response is received. Let's get
+		// response body to return as response to this read command request.
+		// Close response body once read from it
+		defer resp.Body.Close()
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return responses, fmt.Errorf("Read command failed. Cmd:%v err:%v \n", req.DeviceResourceName, err)
+		}
+		driver.logger.Debugf("Resp Body=%s", body)
```
Comment (lenny-intel) [inline driver/restdriver.go:290]: Please remove as it will be way to mush data when the type in Binary or Object.

File: driver/restdriver.go [hunk 19]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
 	return handler.Start()
 }
 
+// Validate data that is being received as part of the http request and response
+func (driver *RestDriver) validateData(resource models.DeviceResource, reading interface{}, valueType string, contentType string) (interface{}, error) {
```
Comment (lenny-intel) [inline driver/restdriver.go:56]: Why are you duplicating the code here: https://github.com/edgexfoundry/device-rest-go/blob/main/driver/resthandler.go#L162
when you could just call it from HandleReadCommands
Yes it will need to be refactored to not be a receiver function so it can be called from both places.

File: cmd/res/profiles/ED_ID1.yaml [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+    isHidden: false
+    description: "used to read DUT Info from the End device"
+    properties:
+        valueType: "String"
+        readWrite: "RW"
+        mediaType: "text/plain"
```
Comment (lenny-intel) [inline cmd/res/profiles/ED_ID1.yaml:23]: `mediaType` only needed and used for Binary and Object types. Please remove from these resources.

File: driver/restdriver.go [hunk 21]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+		reqParam := fmt.Sprint(req.Attributes[URLRawQuery])
+
+		// Form URI from the end device parameters and request parameters and
+		// query parameters. Omit uri prefix if it is empty
+		if uriPrefix != "" {
+			uri = "http://" + address + ":" + port + "/" + uriPrefix + "/" + req.DeviceResourceName + "?" + reqParam + ""
+		} else {
+			uri = "http://" + address + ":" + port + "/" + req.DeviceResourceName + "?" + reqParam + ""
+		}
+		driver.logger.Debugf("uri = %v", uri)
```
Comment (lenny-intel) [inline driver/restdriver.go:255]: This is the one debug statement need. the remaining can be removed.
```suggestion
		driver.logger.Debugf("Sending REST Get command to uri = %v", uri)
```

File: driver/restdriver.go [hunk 22]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+		return fmt.Errorf("ED_PORT not found")
+	}
+	driver.logger.Debugf("ED_PORT:%v", port)
+
+	// Get End devie URI prefix, this can be empty
+	uriPrefix, ok := protocolParams[ED_URI_PREFIX]
+	if !ok {
+		return fmt.Errorf("ED_URI_PREFIX not found")
+	}
+	driver.logger.Debugf("ED_URI_PREFIX:%v", uriPrefix)
```
Comment (lenny-intel) [inline driver/restdriver.go:364]: This is duplicated code. Please create helper function w/o the debug logs.

File: driver/restdriver.go [hunk 23]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+		driver.logger.Debugf("URLRawQuery :%v", req.Attributes[URLRawQuery])
+		reqParam := fmt.Sprint(req.Attributes[URLRawQuery])
+
+		// Form URI from the end device parameters and request parameters and
+		// query parameters. Omit uri prefix if it is empty
+		if uriPrefix != "" {
+			uri = "http://" + address + ":" + port + "/" + uriPrefix + "/" + req.DeviceResourceName + "?" + reqParam + ""
+		} else {
+			uri = "http://" + address + ":" + port + "/" + req.DeviceResourceName + "?" + reqParam + ""
+		}
```
Comment (lenny-intel) [inline driver/restdriver.go:394]: Use fmt.Sprintf to build the uri's

File: driver/restdriver.go [hunk 24]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+
+		// Now we have created http PUT request instance with uri, and payload. This
+		// is enough to initiate PUT request to end device.
+		// First create new http client and initiate PUT request
+		client := &http.Client{}
+
+		resp, err := client.Do(request)
+		if err != nil {
+			// handle error
+			return fmt.Errorf("PUT request failed")
```
Comment (lenny-intel) [inline driver/restdriver.go:453]:
```suggestion
			return fmt.Errorf("PUT request failed to uri = %s", uri)
```

File: driver/restdriver.go [hunk 25]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+		val, err = driver.validateData(deviceResource, reading, deviceResource.Properties.ValueType, contentType)
+		if err != nil {
+			return nil, fmt.Errorf("Recevice response data is not valid")
+		}
+
+		// Now, we have valid response data. This needs to be sent as response to the read command request. Create a CommandValue according to the data type
+		result, err = dsModels.NewCommandValue(deviceResource.Name, deviceResource.Properties.ValueType, val)
+		if err != nil {
+			return nil, err
+		}
```
Comment (lenny-intel) [inline driver/restdriver.go:314]: remove and use new `newCommandValue from restHandler.go

File: driver/restdriver.go [hunk 26]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+	}
+	driver.logger.Debugf("ED_URI_PREFIX:%v", uriPrefix)
+
+	for i, req := range reqs {
+		// First get device resource instance, needed during validation of the
+		// data received in the write command request
+		// RunningService returns the Service instance which is running
+		// service.DeviceResource retrieves the specific DeviceResource instance
+		// from cache according to the Device name and Device Resource name
+		resourceName := req.DeviceResourceName
```
Comment (lenny-intel) [inline driver/restdriver.go:372]: Why create variable when you can just use req.DeviceResourceName?

File: driver/restdriver.go [hunk 27]
```diff
@@ -37,19 +52,422 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+
+		default:
+			return fmt.Errorf("Unsupported value type: %v", valueType)
+		}
+
+		// Now we have created http PUT request instance with uri, and payload. This
+		// is enough to initiate PUT request to end device.
+		// First create new http client and initiate PUT request
+		client := &http.Client{}
+
```
Comment (lenny-intel) [inline driver/restdriver.go:449]: Add Debug logging ("Send PUT command to %s", uri)

File: cmd/res/profiles/ED_ID1.yaml [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+  - name: jsonRes
+    isHidden: false
+    description: "Test resource 1 Message"
+    properties:
+      valueType: "Object"
+      readWrite: "RW"
+      mediaType: "application/json"
+  - name: stringRes
+    isHidden: false
+    description: "used to read DUT Info from the End device"
```
Comment (lenny-intel) [inline cmd/res/profiles/ED_ID1.yaml:19]:
```suggestion
    description: "used to read device Info from the End device"
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]
Thank you for your contribution. Sorry it took till now to get around to reviewing.

**Comment by sudhamani-hcl** [comment]
Hi @lenny-intel ,

I have modified code for all above suggestions.
Could you please review and let me know your comments.

Thanks,
Sudhamani

File: cmd/res/devices/sample-devices.toml [hunk 29]
```diff
@@ -20,3 +20,18 @@
...
+[[DeviceList]]
+  Name = "2way-rest-device"
+  ProfileName = "sample-2way-rest-device"
+  Description = "RESTful Device that sends data"
+  Labels = [ "rest", "2way-rest-device" ]
+  [DeviceList.Protocols]
+    [DeviceList.Protocols.REST]
+	  Host = "127.0.0.1"
+	  Port = "5000"
+	  #Path = ""
```
Comment (lenny-intel) [inline cmd/res/devices/sample-devices.toml:32]: remove commented out line

**Comment by lenny-intel** [comment]
> I have modified code for all above suggestions.
> Could you please review and let me know your comments.

@sudhamani-hcl , something wrong with this PR now. Many existing files are shown as empty.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: cmd/res/devices/sample-devices.toml (hunk 29 above)
Comment (sudhamani-hcl) [reply cmd/res/devices/sample-devices.toml]: Removed

**Comment by sudhamani-hcl** [comment]
@lenny-intel I have resolved the empty files issue. Please review and let me know your comments

File: driver/restdriver.go [hunk 30]
```diff
@@ -38,18 +59,261 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+	var val interface{}
+	var result = &dsModels.CommandValue{}
+	var uri string
+	var protocolParams RestProtocolParams
+	responses = make([]*dsModels.CommandValue, len(reqs))
+
+	// To send request to any end device, first we need to know end device details.
+	// Such as end device IP address, port number on which REST server is running etc.
+	// First get all these details from the device file
+	err = getDeviceParameters(&protocolParams, protocols)
```
Comment (lenny-intel) [inline driver/restdriver.go:74]: function should return `RestProtocolParams`
```suggestion
	protocolParams , err = getDeviceParameters(protocols)
```

File: driver/restdriver.go [hunk 31]
```diff
@@ -38,18 +59,261 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+		if resp.StatusCode > 299 {
+			return fmt.Errorf("PUT request failed with status code: %v", resp.StatusCode)
+		}
+	}
+
+	return nil
+}
+
+// Check for the existance of device parameters in the device file and get them
+func getDeviceParameters(restDeviceProtocolParams *RestProtocolParams, protocols map[string]models.ProtocolProperties) error {
```
Comment (lenny-intel) [inline driver/restdriver.go:292]:
```suggestion
func getDeviceParameters(protocols map[string]models.ProtocolProperties) (RestProtocolParams, error) {
   var restDeviceProtocolParams RestProtocolParams
```

File: driver/restdriver.go [hunk 32]
```diff
@@ -38,18 +59,261 @@ func (driver *RestDriver) Initialize(logger logger.LoggingClient, asyncValues ch
...
+		return fmt.Errorf("RESTPort not found")
+	}
+
+	// Get end device URI prefix, This parameter will be empty for the end devices
+	// which do not have any prefix
+	restDeviceProtocolParams.path, ok = protocolParams[RESTPath]
+	if !ok {
+		return fmt.Errorf("RESTPath not found")
+	}
+	return nil
```
Comment (lenny-intel) [inline driver/restdriver.go:316]:
```suggestion
	return restDeviceProtocolParams, nil
```

File: README.md [hunk 33]
```diff
@@ -138,9 +137,121 @@ The best way to test this service with simulated data is to use **PostMan** to s
...
+
+### Device Profile
+
+This section shows sample device profile defined for REST client functionality.
+
+- [**sample-2way-rest-device**](cmd/res/profiles/sample-2way-rest-device.yaml)
+
+### Testing/Simulation of REST client
+
+This device service supports commanding functionality for the data types as shown in below table.
```
Comment (lenny-intel) [inline README.md:177]:
```suggestion
This device service supports commanding functionality with a sample profile for the data types as shown in below table.
```

File: README.md [hunk 34]
```diff
@@ -138,9 +137,121 @@ The best way to test this service with simulated data is to use **PostMan** to s
...
+| Uint32	| :heavy_check_mark: | :heavy_check_mark: |
+| Uint64	| :heavy_check_mark: | :heavy_check_mark: |
+| Int8	    | :heavy_check_mark: | :heavy_check_mark: |
+| Int16	    | :heavy_check_mark: | :heavy_check_mark: |
+| Int32	    | :heavy_check_mark: | :heavy_check_mark: |
+| Int64	    | :heavy_check_mark: | :heavy_check_mark: |
+| Float32   | :heavy_check_mark: | :heavy_check_mark: |
+| Float64   | :heavy_check_mark: | :heavy_check_mark: |
+
+Using `curl` command-line utility or `PostMan` we can test commanding functionaity of the REST device service.
```
Comment (lenny-intel) [inline README.md:196]: Please explain how to use these tools to provide the 2way rest device that is responding to the GET and PUT requests

File: README.md [hunk 35]
```diff
@@ -27,9 +27,10 @@ make docker-nats
 The locally built Docker image can then be used in place of the published Docker image in your compose file.
 See [Compose Builder](https://github.com/edgexfoundry/edgex-compose/tree/main/compose-builder#gen) `nat-bus` option to generate compose file for NATS and local dev images.
 
-## REST Endpoints
+## REST Server
```
Comment (lenny-intel) [inline README.md:30]: Please revert this change. Original name is correct.

File: README.md [hunk 36]
```diff
@@ -3,15 +3,15 @@
 
 EdgeX device service for REST protocol
 
-This device service provides easy way for 3'rd party applications, such as Point of Sale, CV Analytics, etc., to push data into EdgeX via the REST protocol. 
-
-The current implementation is meant for one-way communication into EdgeX via async readings. If future use cases determine a need for`commanding`, i.e. two-communication, it can be added then.
+This device service supports two-way communication. REST server provides easy way for 3'rd party applications, such as Point of Sale, CV Analytics, etc., to push data into EdgeX via the REST protocol. REST client allows EdgeX to send GET and PUT request to end device.
```
Comment (lenny-intel) [inline README.md:6]:
```suggestion
This device service supports two-way communication for commanding. REST endpoints provides easy way for 3'rd party applications, such as Point of Sale, CV Analytics, etc., to push async data into EdgeX via the REST protocol. Commands allows EdgeX to send GET and PUT request to end device. 
```

File: README.md [hunk 37]
```diff
@@ -27,9 +27,10 @@ make docker-nats
 The locally built Docker image can then be used in place of the published Docker image in your compose file.
 See [Compose Builder](https://github.com/edgexfoundry/edgex-compose/tree/main/compose-builder#gen) `nat-bus` option to generate compose file for NATS and local dev images.
 
-## REST Endpoints
+## REST Server
 
 This device service creates the additional parametrized `REST` endpoint:
+End device will be pushing data to these endpoints. For this end device will be running with REST client.
```
Comment (lenny-intel) [inline README.md:33]:
```suggestion
End device will be pushing async data to these endpoints.
```

File: README.md (hunk 35 above)
Comment (lenny-intel) [reply README.md:30]: Create two sections and this. Existing one is Async and new one is Commands. In Commands device service endpoint is `/api/v2/device/name/{deviceName}/{resourceName}`. Add text about GET and PUT to these endpoints and how the service translates those to GET and PUT to the end REST device.

File: README.md [hunk 38]
```diff
@@ -74,17 +75,15 @@ The `DeviceList` configuration is standard except that the `DeviceList.Protocols
...
 ```
 
-## Device Profile
+### Device Profile
 
 As with all device services the `device profile` is where the **Device Name**, **Device Resources** and **Device Commands** are define. The parameterized REST endpoint described above references these definitions. Each `Device` has it's own device profile. There are three sample device profiles that define the devices referenced in the above sample configuration.
 
 - [**sample-image-device**](cmd/res/profiles/sample-image-device.yaml)
 - [**sample-json-device**](cmd/res/profiles/sample-json-device.yaml)
 - [**sample-numeric-device**](cmd/res/profiles/sample-numeric-device.yaml)
```
Comment (lenny-intel) [inline README.md:84]: Add the new profile here.

File: README.md [hunk 39]
```diff
@@ -74,17 +75,15 @@ The `DeviceList` configuration is standard except that the `DeviceList.Protocols
...
 
-> *Note: The`isHidden` field is set to true since this device service does not support Commanding.* 
-
-## Testing/Simulation
+### Testing/Simulation of REST server
```
Comment (lenny-intel) [inline README.md:86]:
```suggestion
### Testing/Simulation
```

Comment (lenny-intel) [reply README.md:86]: Create subsections for Async (existing) and Commands (new content)

File: README.md [hunk 40]
```diff
@@ -138,9 +137,121 @@ The best way to test this service with simulated data is to use **PostMan** to s
     500.568
     ```
 
+## REST Client
```
Comment (lenny-intel) [inline README.md:140]:
```suggestion
## Commands
```

File: README.md [hunk 41]
```diff
@@ -138,9 +137,121 @@ The best way to test this service with simulated data is to use **PostMan** to s
...
+	  Path = "api"
+  [[DeviceList.AutoEvents]]
+    Interval = "20s"
+    OnChange = false
+    SourceName = "json"
+```
+
+### Device Profile
+
+This section shows sample device profile defined for REST client functionality.
```
Comment (lenny-intel) [inline README.md:171]:
```suggestion
This section shows sample device profile defined for 2 way REST device functionality.
```

File: README.md [hunk 42]
```diff
@@ -138,9 +137,121 @@ The best way to test this service with simulated data is to use **PostMan** to s
...
+
+- [**sample-2way-rest-device**](cmd/res/profiles/sample-2way-rest-device.yaml)
+
+### Testing/Simulation of REST client
```
Comment (lenny-intel) [inline README.md:175]:
```suggestion
### Testing/Simulation Commands
```

File: README.md [hunk 43]
```diff
@@ -138,9 +137,121 @@ The best way to test this service with simulated data is to use **PostMan** to s
...
+| Int8	    | :heavy_check_mark: | :heavy_check_mark: |
+| Int16	    | :heavy_check_mark: | :heavy_check_mark: |
+| Int32	    | :heavy_check_mark: | :heavy_check_mark: |
+| Int64	    | :heavy_check_mark: | :heavy_check_mark: |
+| Float32   | :heavy_check_mark: | :heavy_check_mark: |
+| Float64   | :heavy_check_mark: | :heavy_check_mark: |
+
+Using `curl` command-line utility or `PostMan` we can test commanding functionaity of the REST device service.
+
+**GET Request**
```
Comment (lenny-intel) [inline README.md:198]:
```suggestion
**GET Command**
```

File: README.md [hunk 44]
```diff
@@ -138,9 +137,121 @@ The best way to test this service with simulated data is to use **PostMan** to s
...
+            "valueType" : "Int8"
+         }
+      ],
+      "sourceName" : "int8"
+   },
+   "statusCode" : 200
+   } 
+```
+
+**PUT Request**
```
Comment (lenny-intel) [inline README.md:233]:
```suggestion
**SET Command**
```

**Review by lenny-intel — CHANGES_REQUESTED** [review]

**Comment by lenny-intel** [comment]
@sudhamani-hcl , one small code change left and some rework of the README. Almost there! :-)

File: README.md (hunk 34 above)
Comment (sudhamani-hcl) [reply README.md:199]: @lenny-intel , We are using nodejs based REST emulator as end devices for testing this. CURL commands for sending GET/PUT request to EdgeX.

Comment (lenny-intel) [reply README.md:199]: > We are using nodejs based REST emulator as end devices for testing this

Great, can you include this in a subfolder with directions on using it?

**Review by lenny-intel — APPROVED** [review]
LGTM

**Comment by lenny-intel** [comment]
@sudhamani-hcl , Thank you very much for your contribution.
